File: src/main/java/com/alan/xflowlayout/XflowLayout.java
package com.alan.xflowlayout;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;


import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Created by alan on  2019/4/9 12:56
 */
@SuppressLint("NewApi")
public class XFlowLayout extends Flowlayout {
    public static final int HORIZONTAL = 10010;
    public static final int VERTICAL = 10011;

    List<String> textLsit = new ArrayList<>();

    List<TextView> textViews = new ArrayList<>();

    List<TextView> selectedTextViews = new ArrayList<>();

    private ItemOnclickListener onItemClickListener;

    private ItemLongpressListener onItemLongPressListener;

    private boolean enableMaxSelect;

    private boolean isRandomTextColor;


    public XFlowLayout(Context context) {
       this(context,null);
    }

    public XFlowLayout(Context context, AttributeSet attrs) {
       super(context,attrs);
       setWillNotDraw(false);
    }



    public XFlowLayout setTexts(String[] texts) {
        ArrayList<String> list = new ArrayList<>();
        for (String text : texts) {
            list.add(text);
        }
        setTexts(list);
        return this;
    }

    public XFlowLayout setTexts(ArrayList<String> texts) {
        textLsit.clear();
        textViews.clear();
        if (selectedTexts == null) selectedTexts = new ArrayList<>();
        if(flag!=3)selectedTexts.clear();

        removeAllViews();

        for (final String text : texts) {
            textLsit.add(text);

            final TextView textView;

            if (isSingleSelect || isMoreSelect) textView = new TextView(getContext());
            else
                textView = (TextView) LayoutInflater.from(getContext()).inflate(R.layout.textview_layout, this, false);
            textViews.add(textView);

            textView.setText(text);
            textView.setGravity(Gravity.CENTER);
            textView.setTextColor(defaultTextColor);
            textView.setPadding(itemContentPadding + 5, itemContentPadding, itemContentPadding + 5, itemContentPadding + 7);

            if (drawableId > 0) {
                textView.setBackgroundResource(drawableId);
            } else {
                if(!isAllAttrsDefault)
                textView.setBackground(defaultDrawable());
                ;
            }

            if (isRandomTextColor) textView.setTextColor(Color.parseColor(getRandColor()));

            textView.setTextSize(textSize);
            textView.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (isSingleSelect) setSingleSelected(text, textView);
                    if (isMoreSelect) moreSelect(text, textView);
                    if (onItemClickListener != null) onItemClickListener.onClick(text, textView);
                }
            });
            textView.setOnLongClickListener(new OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    if (onItemLongPressListener != null)
                        onItemLongPressListener.onLongClick(text, textView);

                    return true;
                }
            });



            addView(textView);
        }
        initSelected();
        return this;
    }

    private void initSelected() {
        if(flag<0)return;
        if(flag==1){
        selectedFlag1();
        }else if(flag==2){
            selectedFlag2();
        }else if(flag==3){
            selectedFlag3();
        }
    }

    private void selectedFlag1(){
        if(index<0 || index>textViews.size()){
            flag=-1;
            throw new IndexOutOfBoundsException("index out of bounds ,please check the index");
        }
        String text=textLsit.get(index);
        TextView textView=textViews.get(index);
        moreSelect(text,textView);
    }

    private void selectedFlag2(){
        if(index<0 || index>textViews.size() || index>textViews.size()-1 || index>end){
            flag=-1;
            throw new IndexOutOfBoundsException("index out of bounds ,please check the index");
        }
        if(selectedTexts==null)selectedTexts=new ArrayList<>();
        if(end >textViews.size()-1) end =textViews.size()-1;
        for (int i =index; i <= end; i++) {
            selectedTexts.add(textLsit.get(i));
            selectedTextViews.add(textViews.get(i));
        }
        for (TextView selectedTextView : selectedTextViews) {
            selectedTextView.setBackground(selectedDrawable());
            selectedTextView.setTextColor(slectedTextColor);
        }
    }

    private void selectedFlag3(){
        for (String selectedText : selectedTexts) {
            for (TextView textView : textViews) {
                if(selectedText.equals(textView.getText().toString().trim())){
                    selectedTextViews.add(textView);
                    break;
                }
            }
        }
        for (TextView selectedTextView : selectedTextViews) {
            selectedTextView.setBackground(selectedDrawable());
            selectedTextView.setTextColor(slectedTextColor);
        }
    }

    private void moreSelect(String text, TextView textView) {
        if (selectedTexts == null) selectedTexts = new ArrayList<>();
        if (!selectedTextViews.contains(textView)) {
            if (enableMaxSelect && selectedTextViews.size() >= maxSelect) return;
            selectedTextViews.add(textView);
            selectedTexts.add(text);
        } else {
            selectedTextViews.remove(textView);
            selectedTexts.remove(text);
            textView.setBackground(defaultDrawable());
            textView.setTextColor(defaultTextColor);
            return;
        }
        for (TextView selectedTextView : selectedTextViews) {
            selectedTextView.setBackground(selectedDrawable());
            selectedTextView.setTextColor(slectedTextColor);
        }
    }

    private XFlowLayout setSingleSelected(String text, TextView textView) {
        if(selectedTexts!=null)selectedTexts.clear();
        if(selectedTextViews!=null)selectedTextViews.clear();
        for (TextView view : textViews) {
            view.setTextColor(defaultTextColor);
            view.setBackground(defaultDrawable());
        }
        textView.setTextColor(slectedTextColor);
        textView.setBackground(selectedDrawable());
        singleTeSelectedText = text;
        return this;
    }

    /*
    * set which selected at the beginging
    *
    * */
    public XFlowLayout setSelected(int index,int length){
        if(index>=0 && length>1){
            this.index=index;
            this.end =length;
            flag=2;
        }else {
            flag=-1;
        throw new IllegalArgumentException(" Index must be greater than or equal to zero and end is greater than 1 ");
        }



        return this;
    }

    /*
     * set which selected at the beginging
     * */
    public XFlowLayout setSelected(int index){
        flag=1;
        this.index=index;
        return this;
    }

    /*
     * set which selected at the beginging
     * texts     those is being selected
     * */
    public XFlowLayout setSelected(String[] texts){
        if(texts!=null && texts.length>0){
            if(selectedTexts==null)selectedTexts=new ArrayList<>();
            for (String text : texts) {
                selectedTexts.add(text);
            }
            flag=3;
        }else {
            flag=-1;
            throw new IllegalArgumentException("texts must not null or texts.end must than 0");
        }
        return this;
    }

    /*
    * custom drawable
    * */
    public XFlowLayout setItemBackGroudDrawable( int drawableId) {
        if (drawableId > 0) this.drawableId = drawableId;

        return this;
    }


    /*
    * item background color of default
    * */
    public XFlowLayout setItemDefaultBackGroundColor(int color) {
        if(isAllAttrsDefault)return this;
        defaultBackGroundColor = color;
        return this;
    }


    /*
    * item background color while selected
    * */
    public XFlowLayout setItemSelectedBackGroundColor(int color) {
        if(isAllAttrsDefault)return this;
        selectedbackGroundColor = color;
        return this;
    }


    /*
    * set whether single select
    * */
    public XFlowLayout setIsSingleSlect(boolean isSingleSelect) {
        if(isAllAttrsDefault || isMoreSelect)return this;
        this.isSingleSelect = isSingleSelect;
        if (isSingleSelect) isMoreSelect = false;
        return this;
    }

    /*
    * text size
    * */
    public XFlowLayout setTextSize(float spValue) {
        if(isAllAttrsDefault)return this;
        textSize = spValue;
        return this;
    }

   /*
   * DefaultTextColor
   * */
    public XFlowLayout setDefaultTextColor(int color) {
        if(isAllAttrsDefault)return this;
        defaultTextColor = color;
        return this;
    }

    /*
    * TextColor when selected
    * */
    public XFlowLayout setSelectTextColor(int selectedColor) {
        if(isAllAttrsDefault)return this;
        slectedTextColor = selectedColor;

        return this;
    }

    /*
    * RadiuSize
    * */
    public XFlowLayout setRadiuSize(float radiuSize) {
        if(isAllAttrsDefault)return this;
        radius = radiuSize;
        return this;
    }

    /*
    * DefaultStokeColor
    * */
    public XFlowLayout setDefaultStokeColor(int defaultColor) {
        if(isAllAttrsDefault)return this;
        defaultStokeColor = defaultColor;
        return this;
    }

    /*
    * Border color when selected
    * */
    public XFlowLayout setSelectedStokeColor(int selectedColor) {
        if(isAllAttrsDefault)return this;
        selectedStokeColor = selectedColor;
        return this;
    }

   /*
   * Border size
   * */
    public XFlowLayout setStokeSize(int stokeSize) {
        if(isAllAttrsDefault)return this;
        stokeWidth = stokeSize;
        return this;
    }

    /*
    * Item horizontal spacing
    * */
    public XFlowLayout setItemInternalHorizontal(int pxValue) {
        if(isAllAttrsDefault)return this;
        itemInternal_horizontal = pxValue;
        return this;
    }

    /*
     * Item vertical spacing
     * */
    public XFlowLayout setItemInternalVertical(int pxValue) {
        if(isAllAttrsDefault)return this;
        itemInternal_vertical = pxValue;
        return this;
    }

    /*
    * texteView index
    * */
    public void delete(int index){
        if(index <0 | index>getChildCount()-1)throw new IndexOutOfBoundsException("index out of bounds ,please check the index");
        removeViewAt(index);
        textViews.remove(index);
        TextView child= (TextView) getChildAt(index);
        selectedTextViews.remove(child);
    }

    /*
     * the View will be remove
     * */
    public void delete(TextView textView){
        if(textView==null)return;
        selectedTextViews.remove(textView);
        textViews.remove(textView);
        removeView(textView);
    }

    /*
    * Orirntation
    * */
    public XFlowLayout setOrirntation(int ortentation) {
        if(isAllAttrsDefault)return this;
        this.orientation = ortentation;
        return this;
    }


    /*
    * Selected text when returning to single selection
    * */
    public String getSingleSelectedText() {

        return singleTeSelectedText;
    }


    /*
    * Selected text when multiple selection
    * */
    public ArrayList<String> getMoreSelectedText() {

        return selectedTexts;
    }

    /*
    * all setting before is invalid,and all setting will be the default
    * */
    public XFlowLayout setAllDefault(boolean isAllAttrsDefault ){
        this.isAllAttrsDefault=isAllAttrsDefault;
        if(isAllAttrsDefault){
            reSet();
        }
        return this;
    }

    private void reSet() {
        orientation =HORIZONTAL;
        itemInternal_horizontal = 20;
        itemInternal_vertical =  20;
        radius = getResources().getDimension(R.dimen.dp_4);
        isSingleSelect =  false;
        isMoreSelect = false;
        maxSelect =  3;
        textSize =  15F;
        itemContentPadding = (int)this.getResources().getDimension(R.dimen.dp_6);
        defaultTextColor =  Color.parseColor("#CCCCCC");
        slectedTextColor =  Color.parseColor("#CCCCCC");
        defaultBackGroundColor = Color.WHITE;
        selectedbackGroundColor = Color.WHITE;
        stokeWidth =  1;
        drawableId=0;
        defaultStokeColor = Color.parseColor("#CCCCCC");
        selectedStokeColor =  Color.RED;
    }

    /*
    * Set whether the color of each text is randomly selected
    * */
    public XFlowLayout setItemRandowColor(boolean isRandomColor) {
        if(isAllAttrsDefault)return this;
        isSingleSelect = false;
        isMoreSelect = false;
        isRandomTextColor = isRandomColor;
        return this;
    }

   /*
   * Inner spacing
   * */
    public XFlowLayout setItemContentPadding(int pxValue) {
        if(isAllAttrsDefault)return this;
        itemContentPadding = pxValue;
        return this;
    }

    /*
    * random color
    * */
    private String getRandColor() {
        String r, g, b;
        Random random = new Random();
        r = Integer.toHexString(random.nextInt(256)).toUpperCase();
        g = Integer.toHexString(random.nextInt(256)).toUpperCase();
        b = Integer.toHexString(random.nextInt(256)).toUpperCase();

        r = r.length() == 1 ? "0" + r : r;
        g = g.length() == 1 ? "0" + g : g;
        b = b.length() == 1 ? "0" + b : b;

        return "#" + r + g + b;

    }

    /*
    * Set whether to choose multiple
    * */
    public XFlowLayout setIsMoreSelected(boolean isMoreSlect) {
        if(isAllAttrsDefault || isSingleSelect)return this;

        isMoreSelect = isMoreSlect;
        if (isMoreSelect) {
            isSingleSelect = false;
        }
        return this;
    }



    public XFlowLayout setOnItemClickListener(ItemOnclickListener itemOnclickListener) {
        onItemClickListener = itemOnclickListener;
        return this;
    }

    public XFlowLayout setOnItemLongPressListener(ItemLongpressListener itemLongPressListener) {
        onItemLongPressListener = itemLongPressListener;
        return this;
    }

    public XFlowLayout setMaxSelect(int maxSize) {
        enableMaxSelect = true;
        maxSelect = maxSize;
        return this;
    }


}

File: src/main/java/com/example/xflowlayout/flowlayout/XFlowLayout.kt
package com.example.xflowlayout.flowlayout

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.LayoutInflater
import android.widget.TextView
import com.example.administrator.test.flowlayout.FlowLayout
import com.example.administrator.test.flowlayout.OnItemClickListener
import com.example.administrator.test.flowlayout.OnItemLongPressListener
import com.example.xflowlayout.R
import java.util.*
import kotlin.collections.ArrayList
/**
 * Created by alan on  2019/3/19 15:09
 */
@SuppressLint("NewApi")
class XFlowLayout : FlowLayout {
    //所有的text
    var textLsit = arrayListOf<String>()
        private set
    //textView
    var textViews = arrayListOf<TextView>()
        private set
    //选中的textView
    var selectedTextViews: ArrayList<TextView>? = null
        private set
    //未选中的textView
    var unSelectedTextViews: ArrayList<TextView>? = null
        private set
    //控件单击监听
    private var onItemClickListener: OnItemClickListener? = null
    //控件长按监听
    private var onItemLongPressListener: OnItemLongPressListener? = null
    //最多可选数
    private var enableMaxSelect = false
    //随机文字颜色
    private var isRandomTextColor = false


    constructor(context: Context) : super(context, null) {
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
    }

    fun setTexts(texts: Array<String>): XFlowLayout {
        val list = ArrayList<String>()
        texts.forEach { s -> list.add(s) }
        setTexts(list)
        return this
    }

    fun setTexts(texts: ArrayList<String>): XFlowLayout {
        textLsit.clear()
        textViews.clear()
        selectedTexts?.clear()
        removeAllViews()

        for (text in texts) {
            textLsit.add(text)

            var textView: TextView?

            if (isSingleSelect || isMoreSelect) textView = TextView(context)
            else textView = LayoutInflater.from(context).inflate(R.layout.textview_layout, this, false) as TextView
            textViews.add(textView!!)

            textView.setText(text)
            textView.gravity = Gravity.CENTER
            textView.setTextColor(defaultTextColor)
            textView.setPadding(itemContentPadding + 5, itemContentPadding, itemContentPadding + 5, itemContentPadding + 7)

            if(drawableId>0){
                textView.setBackgroundResource(drawableId)
            }else{
                textView.background = defaultDrawable()
            }

            if (isRandomTextColor) textView.setTextColor(Color.parseColor(getRandColor()))

            textView.setTextSize(textSize)
            textView.setOnClickListener {
                if (isSingleSelect) setSingleSelected(text, textView);
                if (isMoreSelect) moreSelect(text, textView)
                onItemClickListener?.onClick(text, textView)

            }
            textView.setOnLongClickListener {
                onItemLongPressListener?.onLongClick(text, textView)
                true
            }


            addView(textView)
        }
        return this
    }

    //单选内部操作
    private fun setSingleSelected(text: String, textView: TextView): XFlowLayout {
        textViews.forEach { t: TextView? ->
            t?.setTextColor(defaultTextColor)
            t?.background = defaultDrawable()
        }
        textView.setTextColor(slectedTextColor)
        textView.background = selectedDrawable()
        singleTeSelectedText = text
        return this
    }

    //item背景Drawable
    fun setItemBackGroudDrawable(drawableId: Int): XFlowLayout {
        if(drawableId>0)this.drawableId=drawableId

        return this
    }

    //itemm默认背景颜色
    fun setItemBackGroundColor(color: Int): XFlowLayout {
        defaultBackGroundColor = color
        return this
    }

    //item被选中时背景颜色
    fun setItemSelectedBackGroundColor(color: Int): XFlowLayout {
        selectedbackGroundColor = color
        return this
    }

    //设置item圆角值
    fun setItemRadius(radius: Float): XFlowLayout {
        this.radius = radius
        return this
    }

    //设置单选
    fun setIsSingleSlect(isSingleSelect: Boolean): XFlowLayout {
        this.isSingleSelect = isSingleSelect
        if (isSingleSelect) isMoreSelect = false
        return this
    }

    //文字大小
    fun setTextSize(spValue: Float): XFlowLayout {
        textSize = spValue
        return this
    }

    //默认文字颜色
    fun setDefaultTextColor(color: Int): XFlowLayout {
        defaultTextColor = color
        return this
    }

    //选中文字颜色
    fun setSelectTextColor(selectedColor: Int): XFlowLayout {
        slectedTextColor = selectedColor

        return this
    }

    //圆角大小
    fun setRadiuSize(radiuSize: Float): XFlowLayout {
        radius = radiuSize

        return this
    }

    //默认边框颜色
    fun setDefaultStokeColor(defaultColor: Int): XFlowLayout {
        defaultStokeColor = defaultColor
        return this
    }

    //选中时边框颜色
    fun setSelectedStokeColor(selectedColor: Int): XFlowLayout {
        selectedStokeColor = selectedColor
        return this
    }

    //边框大小
    fun setStokeSize(stokeSize: Int): XFlowLayout {
        stokeWidth = stokeSize
        return this
    }

    //item水平方向的间隙
    fun setItemInternalHorizontal(pxValue: Int): XFlowLayout {
        itemInternal_horizontal = pxValue
        return this
    }

    //item垂直方向的间隙
    fun setItemInternalVertical(pxValue: Int): XFlowLayout {
        itemInternal_vertical = pxValue
        return this
    }

    //布局方向
    fun setOrirntation(ortentation: Int): XFlowLayout {
        this.orientation = ortentation
        return this
    }

    //单选text
    fun getSingleSelectedText(): String {

        return singleTeSelectedText
    }

    //多选text
    fun getMoreSelectedText(): ArrayList<String>? {

        return selectedTexts
    }

    //设置是否每个text的颜色随机选定
    fun setItemRandowColor(isRandomColor: Boolean): XFlowLayout {
        isSingleSelect = false;
        isMoreSelect = false
        isRandomTextColor = isRandomColor
        return this
    }

    //设置是否每个text的颜色随机选定
    fun setItemContentPadding(pxValue: Int): XFlowLayout {
        itemContentPadding = pxValue
        return this
    }

    //随机颜色
    fun getRandColor(): String {
        var R: String
        var G: String
        var B: String
        val random = Random()
        R = Integer.toHexString(random.nextInt(256)).toUpperCase()
        G = Integer.toHexString(random.nextInt(256)).toUpperCase()
        B = Integer.toHexString(random.nextInt(256)).toUpperCase()
        R = if (R.length == 1) "0$R" else R
        G = if (G.length == 1) "0$G" else G
        B = if (B.length == 1) "0$B" else B
        return "#" + R + G + B
    }

    //设置是否多选
    fun setIsMoreSelected(isMoreSlect: Boolean): XFlowLayout {
        isMoreSelect = isMoreSlect
        if (isMoreSelect) isSingleSelect = false
        return this
    }

    private fun moreSelect(text: String, textView: TextView) {
        if (selectedTextViews == null) selectedTextViews = ArrayList()
        if (unSelectedTextViews == null) unSelectedTextViews = arrayListOf()
        if (selectedTexts == null) selectedTexts = arrayListOf()
        if (!selectedTextViews!!.contains(textView)) {
            if (enableMaxSelect && selectedTextViews!!.size >= maxSelect) return
            selectedTextViews?.add(textView)
            selectedTexts?.add(text)
        } else {
            selectedTextViews?.remove(textView)
            selectedTexts?.remove(text)
            textView?.background = defaultDrawable()
            return
        }
        selectedTextViews?.forEach { t: TextView? ->
            t?.background = selectedDrawable()
        }
    }

    fun setOnItemClickListener(itemOnclickListener: OnItemClickListener): XFlowLayout {
        onItemClickListener = itemOnclickListener
        return this
    }

    fun setOnItemLongPressListener(itemLongPressListener: OnItemLongPressListener): XFlowLayout {
        onItemLongPressListener = itemLongPressListener
        return this
    }

    fun setMaxSelect(maxSize: Int): XFlowLayout {
        enableMaxSelect = true
        maxSelect = maxSize
        return this
    }

    companion object {
        @JvmField
        val HORIZONTAL = 0   //水平布局方向
        @JvmField
        val VERTICAL = 1    //垂直布局方向
    }

}